### PR TITLE
Add table aws_ecrpublic_repository. Closes #334, Closes #342

### LINKED ISSUE #334: `aws_ecr_repository` table not storing `Public` repositories
**Describe the bug**
Public repositories are not being stored in `aws_ecr_repository` table. Only storing `Private` repositories. Need to validate below use case:- 
Ensure there are no ECR repositories set as Public

**Steampipe version (`steampipe -v`)**
Example: v0.4.0

**Plugin version (`steampipe plugin list`)**
aws: v0.13.0

**To reproduce**
Create a repository as Public and check.

**Expected behavior**
It should insert that record in the table.

**Additional context**
image
image

```
> select * from aws_ecr_repository;
+-----------------+--------------+---------------------------------------------------------+--------------------------------------------------------+---------------------+--------------------
| repository_name | registry_id  | arn                                                     | repository_uri                                         | created_at          | image_tag_mutabilit
+-----------------+--------------+---------------------------------------------------------+--------------------------------------------------------+---------------------+--------------------
| testing         | 632902152528 | arn:aws:ecr:ap-south-1:632902152528:repository/testing  | 632902152528.dkr.ecr.ap-south-1.amazonaws.com/testing  | 2021-04-21 12:17:13 | MUTABLE            
| dfgdfdfg        | 632902152528 | arn:aws:ecr:ap-south-1:632902152528:repository/dfgdfdfg | 632902152528.dkr.ecr.ap-south-1.amazonaws.com/dfgdfdfg | 2021-04-21 12:24:57 | MUTABLE            
+-----------------+--------------+------------------------------------------------------

### LINKED ISSUE #342: Add table aws_ecrpublic_repository
**References**
Add any related links that will help us understand the resource, including vendor documentation, related GitHub issues, and Go SDK documentation.

### PR DESCRIPTION
**Pull request by bigdatasourav**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
SETUP: tests/aws_ecrpublic_repository []

PRETEST: tests/aws_ecrpublic_repository

TEST: tests/aws_ecrpublic_repository
Running terraform
aws_ecrpublic_repository.named_test_resource: Creating...
aws_ecrpublic_repository.named_test_resource: Creation complete after 9s [id=turbottest67853]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

account_id = "013122550996"
aws_partition = "aws"
aws_region = "us-east-1"
registry_id = "013122550996"
resource_aka = "arn:aws:ecr-public::013122550996:repository/turbottest67853"
resource_id = "turbottest67853"
resource_name = "turbottest67853"

Running SQL query: test-get-query.sql
[
  {
    "akas": [
      "arn:aws:ecr-public::013122550996:repository/turbottest67853"
    ],
    "partition": "aws",
    "region": "us-east-1",
    "repository_name": "turbottest67853"
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "akas": [
      "arn:aws:ecr-public::013122550996:repository/turbottest67853"
    ],
    "repository_name": "turbottest67853",
    "title": "turbottest67853"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "arn": "arn:aws:ecr-public::013122550996:repository/turbottest67853",
    "partition": "aws",
    "region": "us-east-1",
    "repository_name": "turbottest67853"
  }
]
✔ PASSED

Running SQL query: test-notfound-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "account_id": "013122550996",
    "akas": [
      "arn:aws:ecr-public::013122550996:repository/turbottest67853"
    ],
    "region": "us-east-1",
    "title": "turbottest67853"
  }
]
✔ PASSED

POSTTEST: tests/aws_ecrpublic_repository

TEARDOWN: tests/aws_ecrpublic_repository

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
select
  repository_name,
  registry_id,
  arn,
  repository_uri,
  created_at,
  region,
  account_id
from
  aws_ecrpublic_repository;
+-----------------+--------------+--------------------------------------------------+------------------------------+---------------------+-----------+--------------+
| repository_name | registry_id  | arn                                              | repository_uri               | created_at          | region    | account_id   |
+-----------------+--------------+--------------------------------------------------+------------------------------+---------------------+-----------+--------------+
| test            | 013122550996 | arn:aws:ecr-public::013122550996:repository/test | public.ecr.aws/i1f4e0n4/test | 2021-04-22 07:44:35 | us-east-1 | 013122550996 |
+-----------------+--------------+--------------------------------------------------+------------------------------+---------------------+-----------+--------------+

select
  title,
  p as principal,
  a as action,
  s ->> 'Effect' as effect,
  s -> 'Condition' as conditions
from
  aws_ecrpublic_repository,
  jsonb_array_elements(policy -> 'Statement') as s,
  jsonb_array_elements_text(s -> 'Principal' -> 'AWS') as p,
  jsonb_array_elements_text(s -> 'Action') as a
where
  s ->> 'Effect' = 'Allow'
  and a in ('*', 'ecr-public:*');
+-------+---------------------------------------------+--------------+--------+------------+
| title | principal                                   | action       | effect | conditions |
+-------+---------------------------------------------+--------------+--------+------------+
| test  | arn:aws:iam::013122550996:role/turbot/admin | ecr-public:* | Allow  | <null>     |
| test  | arn:aws:iam::013122550996:user/sourav       | ecr-public:* | Allow  | <null>     |
| test  | arn:aws:iam::013122550996:role/turbot/admin | *            | Allow  | <null>     |
| test  | arn:aws:iam::013122550996:user/sourav       | *            | Allow  | <null>     |
+-------+---------------------------------------------+--------------+--------+------------+
```
</details>

### CODE REVIEW

File: aws/service.go [hunk 1]
```diff
@@ -308,6 +309,28 @@ func EcrService(ctx context.Context, d *plugin.QueryData, region string) (*ecr.E
 	return svc, nil
 }
 
+// EcrpublicService returns the service connection for AWS ECRPublic service
+func EcrpublicService(ctx context.Context, d *plugin.QueryData, region string) (*ecrpublic.ECRPublic, error) {
```
Comment (Subhajit97) [inline aws/service.go:313]:
```suggestion
func EcrPublicService(ctx context.Context, d *plugin.QueryData, region string) (*ecrpublic.ECRPublic, error) {
```

File: aws/service.go [hunk 2]
```diff
@@ -308,6 +309,28 @@ func EcrService(ctx context.Context, d *plugin.QueryData, region string) (*ecr.E
 	return svc, nil
 }
 
+// EcrpublicService returns the service connection for AWS ECRPublic service
```
Comment (Subhajit97) [inline aws/service.go:312]:
```suggestion
// EcrpublicService returns the service connection for AWS ECR Public service
```

File: aws/table_ aws_ecrpublic_repository.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+	"github.com/turbot/steampipe-plugin-sdk/plugin/transform"
+)
+
+//// TABLE DEFINITION
+
+func tableAwsEcrpublicRepository(_ context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "aws_ecrpublic_repository",
+		Description: "AWS ECRPublic Repository",
```
Comment (Subhajit97) [inline aws/table_ aws_ecrpublic_repository.go:19]: https://docs.aws.amazon.com/AmazonECR/latest/public/public-repositories.html
```suggestion
		Description: "AWS ECR Public Repository",
```

File: aws/table_ aws_ecrpublic_repository.go [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+		GetMatrixItem: BuildRegionList,
+		Columns: awsRegionalColumns([]*plugin.Column{
+			{
+				Name:        "repository_name",
+				Description: "The name of the repository.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "registry_id",
+				Description: "The AWS account ID associated with the public registry that contains the repository",
```
Comment (Subhajit97) [inline aws/table_ aws_ecrpublic_repository.go:37]:
```suggestion
				Description: "The AWS account ID associated with the public registry that contains the repository.",
```

File: aws/table_ aws_ecrpublic_repository.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+			{
+				Name:        "policy",
+				Description: "The JSON repository policy text associated with the repository.",
+				Hydrate:     getAwsEcrpublicRepositoryPolicy,
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("PolicyText"),
+			},
+			{
+				Name:        "tags_src",
+				Description: "A list of tags assigned to the Repository.",
```
Comment (Subhajit97) [inline aws/table_ aws_ecrpublic_repository.go:71]:
```suggestion
				Description: "A list of tags assigned to the repository.",
```

File: aws/table_ aws_ecrpublic_repository.go [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     listAwsEcrpublicRepositoryTags,
+				Transform:   transform.FromField("Tags"),
+			},
+
+			// Standard columns for all tables
```
Comment (Subhajit97) [inline aws/table_ aws_ecrpublic_repository.go:77]:
```suggestion
			// Steampipe standard columns
```

File: aws/table_ aws_ecrpublic_repository.go [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+				Description: resourceInterfaceDescription("title"),
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("RepositoryName"),
+			},
+			{
+				Name:        "akas",
+				Description: resourceInterfaceDescription("akas"),
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("RepositoryArn").Transform(arnToAkas),
+			},
```
Comment (Subhajit97) [inline aws/table_ aws_ecrpublic_repository.go:96]: Order should be `title`, `tags` and `akas`.

File: aws/table_ aws_ecrpublic_repository.go [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+			return !isLast
+		},
+	)
+
+	return nil, err
+}
+
+////  HYDRATE FUNCTIONS
+
+func getAwsEcrpublicRepositories(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```
Comment (Subhajit97) [inline aws/table_ aws_ecrpublic_repository.go:135]:
```suggestion
func getAwsEcrpublicRepository(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```

File: aws/table_ aws_ecrpublic_repository.go [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+	logger := plugin.Logger(ctx)
+	logger.Trace("getAwsEcrpublicRepositories")
+
+	// TODO put me in helper function
+	var region string
+	matrixRegion := plugin.GetMatrixItem(ctx)[matrixKeyRegion]
+	if matrixRegion != nil {
+		region = matrixRegion.(string)
+	}
+	var name string
```
Comment (Subhajit97) [inline aws/table_ aws_ecrpublic_repository.go:145]:
```suggestion

	var name string
```

File: aws/table_ aws_ecrpublic_repository.go [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+		RepositoryNames: []*string{aws.String(name)},
+	}
+
+	// Get call
+	data, err := svc.DescribeRepositories(params)
+	if err != nil {
+		logger.Debug("getAwsEcrpublicRepositories", "ERROR", err)
+		return nil, err
+	}
+	return data.Repositories[0], nil
```
Comment (Subhajit97) [inline aws/table_ aws_ecrpublic_repository.go:169]: can you please add a if block to check the length of the response?

File: aws/table_ aws_ecrpublic_repository.go [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+
+	return op, nil
+}
+
+//// TRANSFORM FUNCTIONS
+
+func ecrpublicTagListToTurbotTags(ctx context.Context, d *transform.TransformData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("ecrpublicTagListToTurbotTags")
+	tags := d.HydrateItem.(*ecrpublic.ListTagsForResourceOutput)
+
```
Comment (Subhajit97) [inline aws/table_ aws_ecrpublic_repository.go:279]:
```suggestion

if tags == nil {
  return nil, nil
}

```

File: aws-test/tests/aws_ecrpublic_repository/test-turbot-query.sql [hunk 12]
```diff
@@ -0,0 +1,3 @@
+select account_id, akas, region, title
+from aws.aws_ecrpublic_repository
+where repository_name = '{{ output.resource_name.value }}';
```
Comment (Subhajit97) [inline aws-test/tests/aws_ecrpublic_repository/test-turbot-query.sql:3]:
```suggestion
where repository_name = '{{ resourceName }}';
```

File: aws-test/tests/aws_ecrpublic_repository/test-notfound-query.sql [hunk 13]
```diff
@@ -0,0 +1,4 @@
+select registry_id, repository_name, arn, region, account_id
+from aws.aws_ecr_repository
+where repository_name = '{{ output.resourceName }}.1';
```
Comment (Subhajit97) [inline aws-test/tests/aws_ecrpublic_repository/test-notfound-query.sql:3]:
```suggestion
where repository_name = '{{ resourceName }}.1';
```

File: aws-test/tests/aws_ecrpublic_repository/test-list-expected.json [hunk 14]
```diff
@@ -0,0 +1,8 @@
+[
+  {
+    "arn": "{{ output.resource_aka.value }}",
+    "partition": "{{ output.aws_partition.value }}",
+    "region": "{{ output.aws_region.value }}",
+    "repository_name": "{{ output.resource_name.value }}"
```
Comment (Subhajit97) [inline aws-test/tests/aws_ecrpublic_repository/test-list-expected.json:6]:
```suggestion
    "repository_name": "{{ resourceName }}"
```

**Review by Subhajit97 — CHANGES_REQUESTED** [review]
@bigdatasourav There are few changes required. Please see comments, thanks!

File: aws-test/tests/aws_ecrpublic_repository/test-notfound-query.sql [hunk 15]
```diff
@@ -0,0 +1,4 @@
+select registry_id, repository_name, arn, region, account_id
+from aws.aws_ecr_repository
```
Comment (khushboo9024) [inline aws-test/tests/aws_ecrpublic_repository/test-notfound-query.sql:2]:
```suggestion
from aws. aws_ecrpublic_repository
```

File: aws/table_ aws_ecrpublic_repository.go [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+	// Build the params
+	params := &ecrpublic.DescribeRepositoriesInput{
+		RepositoryNames: []*string{aws.String(name)},
+	}
+
+	// Get call
+	data, err := svc.DescribeRepositories(params)
+	if err != nil {
+		logger.Debug("getAwsEcrpublicRepositories", "ERROR", err)
```
Comment (Subhajit97) [inline aws/table_ aws_ecrpublic_repository.go:167]:
```suggestion
		logger.Debug("getAwsEcrpublicRepository", "ERROR", err)
```

File: aws/table_ aws_ecrpublic_repository.go [hunk 17]
```diff
@@ -0,0 +1,299 @@
...
+		return nil, err
+	}
+	if len(data.Repositories) > 0 {
```
Comment (Subhajit97) [inline aws/table_ aws_ecrpublic_repository.go:170]:
```suggestion
	if data.Repositories != nil && len(data.Repositories) > 0 {
```

File: aws/table_ aws_ecrpublic_repository.go [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+	if matrixRegion != nil {
+		region = matrixRegion.(string)
+	}
+
+	var name string
+	if h.Item != nil {
+		name = *h.Item.(*ecrpublic.Repository).RepositoryName
+	} else {
+		name = d.KeyColumnQuals["repository_name"].GetStringValue()
+	}
```
Comment (Subhajit97) [inline aws/table_ aws_ecrpublic_repository.go:151]: Is this check required?

File: aws/table_ aws_ecrpublic_repository.go [hunk 19]
```diff
@@ -0,0 +1,299 @@
+package aws
```
Comment (Subhajit97) [inline aws/table_ aws_ecrpublic_repository.go:1]: Fix the file name. There is an space in after `table_`

**Review by Subhajit97 — CHANGES_REQUESTED** [review]
@bigdatasourav There are few more changes required. Please see comments, thanks!

File: aws/table_aws_ecrpublic_repository.go [hunk 20]
```diff
@@ -0,0 +1,294 @@
...
+			return !isLast
+		},
+	)
+
+	return nil, err
+}
+
+////  HYDRATE FUNCTIONS
+
+func getAwsEcrpublicRepository(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```
Comment (Subhajit97) [inline aws/table_aws_ecrpublic_repository.go:135]:
```suggestion
func getAwsEcrpublicRepository(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```

**Review by Subhajit97 — APPROVED** [review]
LGTM

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by khushboo9024 — APPROVED** [review]

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**